File: src/main/java/dev/klomptech/jbark/parser/collection/StashParser.java
package dev.klomptech.jbark.parser.collection;

import dev.klomptech.jbark.config.ConfigLoader;
import dev.klomptech.jbark.errors.BarkError;
import dev.klomptech.jbark.lexer.Token;
import dev.klomptech.jbark.lexer.TokenType;
import dev.klomptech.jbark.parser.AstNode;
import dev.klomptech.jbark.parser.Keywords;
import dev.klomptech.jbark.parser.ParseExpression;
import dev.klomptech.jbark.parser.Parser;
import dev.klomptech.jbark.parser.expression.ValueParser;
import java.util.ArrayList;
import java.util.List;

public class StashParser {

  private final Parser parser;

  public StashParser(final Parser parser) {
    this.parser = parser;
  }

  // Fill or clear a registered stash (my list type) on the line
  public AstNode parseStashLine() throws BarkError {
    int line = parser.peek().line();
    String stash = findStashOnLine();
    if (stash == null) {
      return null;
    }
    // Make sure the stash is registered
    ConfigLoader.requireStash(stash, line);

    // Look for the holds keyword
    if (parser.lineHasWord(Keywords.HOLDS_WORDS)) {
      List<ParseExpression> items = parseItemsAfterHolds();
      if (items.isEmpty()) {
        throw new BarkError(line, "The jar is open but nothing went inside. How dare you?");
      }
      parser.consumeUntilLineEnd();
      return new AstNode.StashInit(stash, items, line);
    }
    // Stash gains one item
    if (parser.lineHasWord(Keywords.STASH_ADD_WORDS) && !lineHasFrom()) {
      ParseExpression item = parseItemAfterStashAdd();
      if (item == null) {
        throw new BarkError(
            line, "You didn't say what will be added to the jar for the dogs to steal later!");
      }
      parser.consumeUntilLineEnd();
      return new AstNode.StashAppend(stash, item, line);
    }
    // Example: "her cookie jar drops first" removes the item in that list slot
    if (parser.lineHasWord(Keywords.DROPS_WORDS)) {
      String slot = findJarSlotOnLine();
      if (slot == null) {
        throw new BarkError(line, "Which item? The dogs need to know which one to steal");
      }
      parser.consumeUntilLineEnd();
      return new AstNode.StashRemove(stash, slot, line);
    }
    if (!lineHasFrom() && !parser.lineHasWord(Keywords.TAKE_FROM_STASH_WORDS)) {
      // slot is first, second, third, last, etc
      String slot = findJarSlotOnLine();
      if (slot != null) {
        ParseExpression newValue = parseNewValueAfterSlot(slot);
        if (newValue == null) {
          return null;
        }
        parser.consumeUntilLineEnd();
        return new AstNode.StashSet(stash, slot, newValue, line);
      }
    }
    // Look for the clear keyword
    if (lineHasNullClear()) {
      parser.consumeUntilLineEnd();
      return new AstNode.StashClear(stash, line);
    }
    return null;
  }

  private String findStashOnLine() {
    // Look for the stash name
    String name =
        ConfigLoader.resolveCollectionFromTokens(
            parser.allTokens(), parser.cursorIndex(), parser.lineEndIndex());
    if (name == null || !ConfigLoader.isStash(name)) {
      return null;
    }
    return name;
  }

  private boolean lineHasFrom() {
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.IDENTIFIER)
          && Keywords.FROM_WORDS.contains(parser.normalise(token.value()))) {
        return true;
      }
    }
    return false;
  }

  private boolean lineHasNullClear() {
    // Look for the clear keyword (which means empty list)
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.IDENTIFIER)
          && Keywords.isNullClearWord(parser.normalise(token.value()))) {
        return true;
      }
    }
    return false;
  }

  // List slot word on the line: first, second, third, last, etc
  private String findJarSlotOnLine() {
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      String word = parser.normalise(token.value());
      if (StashSpots.isStashSpot(word)) {
        return word;
      }
    }
    return null;
  }

  // Value after the slot word: a plain number, a string, or a full expression
  private ParseExpression parseNewValueAfterSlot(final String slot) throws BarkError {
    int slotAt = -1;
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.IDENTIFIER) && slot.equals(parser.normalise(token.value()))) {
        slotAt = offset;
        break;
      }
    }
    if (slotAt < 0) {
      return null;
    }
    int valueStart = -1;
    for (int offset = slotAt + 1; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.IDENTIFIER)
          && Keywords.isAssignSubjectGlue(parser.normalise(token.value()))) {
        valueStart = offset + 1;
        break;
      }
    }
    if (valueStart < 0) {
      return null;
    }
    return new ValueParser(parser).parsePart(valueStart, parser.countTokensAhead(0));
  }

  private List<ParseExpression> parseItemsAfterHolds() {
    int holdsAt = -1;
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      // Look for the holds keyword
      if (token.is(TokenType.IDENTIFIER)
          && Keywords.HOLDS_WORDS.contains(parser.normalise(token.value()))) {
        holdsAt = offset;
        break;
      }
    }
    List<ParseExpression> items = new ArrayList<>();
    if (holdsAt < 0) {
      return items;
    }
    for (int offset = holdsAt + 1; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      // Skip commas and list and words
      if (token.is(TokenType.COMMA)) {
        continue;
      }
      // Skip list and words
      if (token.is(TokenType.IDENTIFIER)
          && Keywords.LIST_AND_WORDS.contains(parser.normalise(token.value()))) {
        continue;
      }
      // Look for the items
      if (token.is(TokenType.STRING)) {
        items.add(new ParseExpression.StringLiteral(token.value()));
        // Look for the numbers
      } else if (token.is(TokenType.NUMBER)) {
        items.add(
            new ParseExpression.NumberLiteral(Double.parseDouble(parser.normalise(token.value()))));
      }
    }
    return items;
  }

  // First string or number after stow/snag/get/collect on this line
  private ParseExpression parseItemAfterStashAdd() {
    int addAt = -1;
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      // Look for the stash add word
      if (token.is(TokenType.IDENTIFIER)
          && Keywords.STASH_ADD_WORDS.contains(parser.normalise(token.value()))) {
        addAt = offset;
        break;
      }
    }
    if (addAt < 0) {
      return null;
    }
    // Look for the item
    for (int offset = addAt + 1; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.STRING)) {
        return new ParseExpression.StringLiteral(token.value());
      }
      if (token.is(TokenType.NUMBER)) {
        return new ParseExpression.NumberLiteral(
            Double.parseDouble(parser.normalise(token.value())));
      }
    }
    return null;
  }
}

File: pbark/pbark/parser/collection/stash_parser.py
from __future__ import annotations

from pbark.config.config_loader import ConfigLoader
from pbark.errors import BarkError
from pbark.lexer import TokenType
from pbark.parser import ast_node as AstNode
from pbark.parser import keywords as Keywords
from pbark.parser.collection import stash_spots as StashSpots
from pbark.parser.expression.value_parser import ValueParser
from pbark.parser.parse_expression import NumberLiteral, StringLiteral


class StashParser:
    def __init__(self, parser) -> None:
        self._parser = parser

    def parse_stash_line(self):
        line = self._parser.peek().line
        stash = self._find_stash_on_line()
        if stash is None:
            return None
        ConfigLoader.require_stash(stash, line)
        if self._parser.line_has_word(Keywords.HOLDS_WORDS):
            items = self._parse_items_after_holds()
            if not items:
                raise BarkError(line, "The jar is open but nothing went inside. How dare you?")
            self._parser.consume_until_line_end()
            return AstNode.StashInit(stash, items, line)
        if self._parser.line_has_word(Keywords.STASH_ADD_WORDS) and not self._line_has_from():
            item = self._parse_item_after_stash_add()
            if item is None:
                raise BarkError(
                    line,
                    "You didn't say what will be added to the jar for the dogs to steal later!",
                )
            self._parser.consume_until_line_end()
            return AstNode.StashAppend(stash, item, line)
        if self._parser.line_has_word(Keywords.DROPS_WORDS):
            slot = self._find_jar_slot_on_line()
            if slot is None:
                raise BarkError(line, "Which item? The dogs need to know which one to steal")
            self._parser.consume_until_line_end()
            return AstNode.StashRemove(stash, slot, line)
        if not self._line_has_from() and not self._parser.line_has_word(Keywords.TAKE_FROM_STASH_WORDS):
            slot = self._find_jar_slot_on_line()
            if slot is not None:
                new_value = self._parse_new_value_after_slot(slot)
                if new_value is None:
                    return None
                self._parser.consume_until_line_end()
                return AstNode.StashSet(stash, slot, new_value, line)
        if self._line_has_null_clear():
            self._parser.consume_until_line_end()
            return AstNode.StashClear(stash, line)
        return None

    def _find_stash_on_line(self) -> str | None:
        name = ConfigLoader.resolve_collection_from_tokens(
            self._parser.all_tokens(), self._parser.cursor_index(), self._parser.line_end_index()
        )
        if name is None or not ConfigLoader.is_stash(name):
            return None
        return name

    def _line_has_from(self) -> bool:
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER) and self._parser.normalise(token.value) in Keywords.FROM_WORDS:
                return True
            offset += 1
        return False

    def _line_has_null_clear(self) -> bool:
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER) and Keywords.is_null_clear_word(
                self._parser.normalise(token.value)
            ):
                return True
            offset += 1
        return False

    def _find_jar_slot_on_line(self) -> str | None:
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if not token.is_(TokenType.IDENTIFIER):
                offset += 1
                continue
            word = self._parser.normalise(token.value)
            if StashSpots.is_stash_spot(word):
                return word
            offset += 1
        return None

    def _parse_new_value_after_slot(self, slot: str):
        slot_at = -1
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER) and slot == self._parser.normalise(token.value):
                slot_at = offset
                break
            offset += 1
        if slot_at < 0:
            return None
        value_start = -1
        off = slot_at + 1
        while not self._parser.is_at_end_or_line_end_offset(off):
            token = self._parser.peek_at(off)
            if token.is_(TokenType.IDENTIFIER) and Keywords.is_assign_subject_glue(
                self._parser.normalise(token.value)
            ):
                value_start = off + 1
                break
            off += 1
        if value_start < 0:
            return None
        return ValueParser(self._parser).parse_part(value_start, self._parser.count_tokens_ahead(0))

    def _parse_items_after_holds(self):
        holds_at = -1
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER) and self._parser.normalise(token.value) in Keywords.HOLDS_WORDS:
                holds_at = offset
                break
            offset += 1
        items = []
        if holds_at < 0:
            return items
        off = holds_at + 1
        while not self._parser.is_at_end_or_line_end_offset(off):
            token = self._parser.peek_at(off)
            if token.is_(TokenType.COMMA):
                off += 1
                continue
            if token.is_(TokenType.IDENTIFIER) and self._parser.normalise(token.value) in Keywords.LIST_AND_WORDS:
                off += 1
                continue
            if token.is_(TokenType.STRING):
                items.append(StringLiteral(token.value))
            elif token.is_(TokenType.NUMBER):
                items.append(NumberLiteral(float(self._parser.normalise(token.value))))
            off += 1
        return items

    def _parse_item_after_stash_add(self):
        add_at = -1
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER) and self._parser.normalise(token.value) in Keywords.STASH_ADD_WORDS:
                add_at = offset
                break
            offset += 1
        if add_at < 0:
            return None
        off = add_at + 1
        while not self._parser.is_at_end_or_line_end_offset(off):
            token = self._parser.peek_at(off)
            if token.is_(TokenType.STRING):
                return StringLiteral(token.value)
            if token.is_(TokenType.NUMBER):
                return NumberLiteral(float(self._parser.normalise(token.value)))
            off += 1
        return None
